### ISSUE TITLE
[Feature]: Support dynamic schema for collection

### ISSUE THREAD

**Issue by Writtic** [milvus-io/milvus#15372]
### Is there an existing issue for this?

- [X] I have searched the existing issues

### Is your feature request related to a problem? Please describe.

Based on https://github.com/milvus-io/milvus/pull/14418 comments, `entityExpiration` can't limit each collection separately unless the variable is on the collection schema. For now, the variable is on the global config, so all collection should follow the `entityExpiration`.

### Describe the solution you'd like.

The solution is that add more collection schema or change dynamically.

**Comment by congqixia** [comment]
It make sense that `entityExpiration` should be an attribute of a collection. Shall we make it mutable, which could be alter by DDL? Or make it immutable, which is determined when collection is created.

**Comment by Writtic** [comment]
Maybe when alter the schema on the production, it might occur unexpected behavior especially on compaction. And altering feature will support to mutate not only `entityExpiration` but also the other variables, right? If it is, Milvus should guarantee some level of integrity after schema changes too, I guess.

**Comment by xiaofan-luan** [comment]
> It make sense that `entityExpiration` should be an attribute of a collection. Shall we make it mutable, which could be alter by DDL? Or make it immutable, which is determined when collection is created.

I would prefer to be mutable. For first step we may only support collection to be altered under released states, thus the integrity should be much easier to maintain.
The second step is to do online attributes change, we can do the alter by segment/channel so that won't affect search/inserts

**Comment by xiaofan-luan** [comment]
> les,

Yep, entityExpirationTime, or TTL(time to live) should be a very beginning to start.

**Comment by Writtic** [comment]
Hi, @xiaofan-luan. I have a question about this issue.

### To support mutable attribute of collection level 
`CollectionSchema` and `CollectionMeta` doesn't seem to be mutable candidates. 
Shall I make another `struct` matched to `CollectionID`? 
Otherwise, just define and implement new task like `UpdateCollectionTask` on DDL request process?

**Comment by xiaofan-luan** [comment]
> Hi, @xiaofan-luan. I have a question about this issue.
> 
> ### To support mutable attribute of collection level
> `CollectionSchema` and `CollectionMeta` doesn't seem to be mutable candidates. Shall I make another `struct` matched to `CollectionID`? Otherwise, just define and implement new task like `UpdateCollectionTask` on DDL request process?

Hi @Writtic 
I would like to introduce another call named AlterCollectionMeta/AlterCollectionSchema and the collection meta itself should not be immutable.

Also to implement this call we has to notify all place maintain collection metacache? I'm not sure where it locates, might be proxy, datanode and querynode

**Comment by congqixia** [comment]
I think this feature is still needed. Let's keep it open.

**Comment by CaptXiong** [comment]
> 

This feature is needed. If anyone working on this new feature?

**Comment by xiaofan-luan** [comment]
Don't think so.
To support dynamic schema, we need huge refactor on execute engine, maybe support struct like map/list would also work?

**Comment by arunkaly** [comment]
Are there any workarounds currently (or in the near future) to address this issue. I can think of options like allowing a rename / alias of the Collection, and that way allow addition of new Schema fields in that operation. This ca be a slow operation that would update all the records with null values for the newly added field. In the execution engine, you treat them as regular fields where some have null values while others don't?
